official/wide_deep/movie_dataset.py

```diff
@@ -0,0 +1,303 @@
+    tf.gfile.Remove(buffer_path)
+
+  tf.logging.info("Constructing {}".format(buffer_path))
+  with tf.python_io.TFRecordWriter(buffer_path) as writer:
+    for row in dataframe.itertuples():
+      i = getattr(row, "Index")
+      features = {}
+      for key in columns:
+        value = getattr(row, key)
+        if isinstance(value, int):
+          features[key] = tf.train.Feature(
+              int64_list=tf.train.Int64List(value=[value]))
+        elif isinstance(value, float):
+          features[key] = tf.train.Feature(
+              float_list=tf.train.FloatList(value=[value]))
+        elif isinstance(value, (six.text_type, six.binary_type)):
+          if not isinstance(value, six.binary_type):
+            value = value.encode("utf-8")
+          features[key] = tf.train.Feature(
+              bytes_list=tf.train.BytesList(value=[value]))
+        else:
+          raise ValueError("Unknown type.")
+      example = tf.train.Example(features=tf.train.Features(feature=features))
+      writer.write(example.SerializeToString())
+      if (i + 1) % 50000 == 0:
+        tf.logging.info(
+            "{}/{} examples written.".format(str(i+1).ljust(8), len(dataframe)))
+
+
+def _deserialize(example_serialized):
+  features = tf.parse_single_example(example_serialized, _FEATURE_MAP)
+  return features, features["rating"]
+
+
+def get_input_fns(data_dir, small=True, repeat=1, batch_size=32):
+  """Construct training and eval input_fns."""
+  ratings, metadata = read_and_process_data(data_dir, small)
+  feature_columns = construct_feature_columns(metadata, ratings)
+
+  def model_column_fn():
+    # There are no wide columns. Only deep columns for now.
+    return [], feature_columns
+
+  movie_columns = ["movieId", "year", "original_language",
+                   "genres_0", "genres_1", "budget"]
+
+  merged_data = pd.merge(ratings, metadata[movie_columns], on="movieId")
+  all_columns = ["userId"] + movie_columns + ["rating"]
+
+  train_df = merged_data.sample(frac=0.8, random_state=0)
+  eval_df = merged_data.drop(train_df.index)
+
+  train_df = train_df.reset_index(drop=True)
+  eval_df = eval_df.reset_index(drop=True)
+
+  tf.logging.info("Training: {} examples".format(len(train_df)))
+  tf.logging.info("Eval:     {} examples".format(len(eval_df)))
+
+  buffer_dir = os.path.join(data_dir, _BUFFER_SUBDIR)
+  tf.gfile.MakeDirs(buffer_dir)
+
+  def _construct_input_fn(dataframe, buffer_name, repeat=1, batch_size=1):
+    """Construct a file backed dataset and input_fn from a dataframe."""
+    buffer_path = os.path.join(buffer_dir, buffer_name)
+    _write_to_buffer(
+        dataframe=dataframe, buffer_path=buffer_path, columns=all_columns)
+
+    def input_fn():
+      dataset = tf.data.TFRecordDataset(buffer_path)
+      dataset = dataset.repeat(repeat)
+      dataset = dataset.map(_deserialize)
+      dataset = dataset.batch(batch_size)
+      return dataset.prefetch(1)
+
+    return input_fn
```

Comment:
Is this _write_to_buffer the magic here? For the recommendation model with numpy array data, should I replace the dataframe with numpy array and perform desalinization on the serialized binary file?